### PR TITLE
calc _blackBoxRate based on real timestamps

### PR DESCRIPTION
**Pull request by mituritsyn**
In some conditions the actual logging speed may differ from that set by the settings, it may be worth minimising such situations at the firmware level, but if we get such data we will be able to adequately handle them.
Issue was reported by @demvlad

[Uploading HQProp.zip…]()

### CODE REVIEW

**Comment by demvlad** [comment]
Yes! You found log records count computing in 3 rows code!
What i think.
I try to use actual log rate for good log file from my other quad. For example, one log has 2002Hz actual rate instead of 2000 from config data. But in this case i see that frequency axis has main ticks labels with fractional (not int) values. May be use actual rates in case of significantly rate difference only (5% for example).
The situations like my FC are rare. Maybe show some warning at the spectrum chart in this case.
I have my realization of resolving this issue what is showed warnings at spectrum chart. But it has bad realization of logs records count computing.  I can add warning show into your code.

File: js/graph_spectrum_calc.js [hunk 1]
```diff
@@ -24,15 +24,14 @@ var GraphSpectrumCalc = GraphSpectrumCalc || {
...
     this._sysConfig = sysConfig;
 
-    var gyroRate = (1000000 / this._sysConfig['looptime']).toFixed(0);
-    this._blackBoxRate = gyroRate * this._sysConfig['frameIntervalPNum'] / this._sysConfig['frameIntervalPDenom'];
-    if (this._sysConfig.pid_process_denom != null) {
-        this._blackBoxRate = this._blackBoxRate / this._sysConfig.pid_process_denom;
-    }
+    let minTime = this._flightLog.getMinTime();
+    let maxTime = this._flightLog.getMaxTime();
+    var allChunks = this._flightLog.getChunksInTimeRange(minTime, maxTime);
```
Comment (haslinghuis) [inline js/graph_spectrum_calc.js:32]:
```suggestion
    const minTime = this._flightLog.getMinTime();
    const maxTime = this._flightLog.getMaxTime();
    const allChunks = this._flightLog.getChunksInTimeRange(minTime, maxTime);
```

**Comment by demvlad** [comment]
or would you mind if I will use your codes fragment in my realization?

**Comment by mituritsyn** [comment]
> or would you mind if I will use your codes fragment in my realization?

I have a feeling that this implementation has significantly increased graphs loading time, I want to test another way to calculate the data rows count, but it is not so elegant and is based on some assumptions that need to be tested as well.

File: js/graph_spectrum_calc.js (hunk 1 above)
Comment (mituritsyn) [reply js/graph_spectrum_calc.js:32]: Is it ok to make the same changes in the remaining functions of this js file in this pr? I also want to do some minor refactoring here, as it is best way to dive into project

**Comment by demvlad** [comment]
> I have a feeling that this implementation has significantly increased graphs loading time, I want to test another way to calculate the data rows count, but it is not so elegant and is based on some assumptions that need to be tested as well.

This is good, if faster solution is exists.
I've notice, that this procedure have some delay for big time interval. There is the time limit 5min flight time in source specter code.
But i think, maybe make this checking once when log is loaded. As we look, this is a bad situation till. Maybe better to informed about it after log loading. And use this data from log class in spectrum chart as ready.

I am not full study till, how works log data parsing code... I have some questions to minmax computing and want to resolve their in future.

File: js/graph_spectrum_calc.js [hunk 2]
```diff
@@ -45,7 +45,7 @@ GraphSpectrumCalc.setOutTime = function(time) {
         this._analyserTimeRange.out = time;
     } else {
         this._analyserTimeRange.out = analyserTimeRange.in + MAX_ANALYSER_LENGTH;
-    }
+    };
```
Comment (haslinghuis) [inline js/graph_spectrum_calc.js:48]: If statements do not need semicolon. See more below.

**Review by haslinghuis — CHANGES_REQUESTED** [review]

**Comment by haslinghuis** [comment]
@mituritsyn please rebase - to resolve conflicts.

**Comment by nerdCopter** [comment]
@demvlad , does this fit into your #711 somehow, or is it a duplication in essence? 
when i attempted to merge locally there is obvious conflict.
This PR will need closure, but unsure if a new "PWA" compatible/rebase version need to be opened or not.

**Pull request closed without merging**